src/Resolver.hack

```diff
@@ -26,25 +26,28 @@ final class Resolver {
         dict<string, mixed> $variables = dict[],
         ?string $operation_name = null,
     ): Awaitable<this::TResponse> {
-        // TODO: catch these errors
-        $source = new \Graphpinator\Source\StringSource($input);
-        $parser = new \Graphpinator\Parser\Parser($source);
-        $request = $parser->parse();
+        $request = null;
+        $errors = vec[];
+
+        try {
+            $request = $this->parseRequest($input);
+        } catch (GraphQL\UserFacingError $e) {
+            $errors[] = $e;
+        }
 
         $ret = shape();
-        $validator = new \Slack\GraphQL\Validation\Validator($this->schema);
-        $errors = $validator->validate($request);
-
-        if (C\is_empty($errors)) {
-            try {
-                list($ret['data'], $errors) = await $this->resolveImpl($request, $variables, $operation_name);
-            } catch (UserFacingError $e) {
-                $errors = vec[$e];
-            } catch (\Throwable $e) {
-                // TODO: This shoud not happen; if it does, it's a bug in the GraphQL framework. Every exception should
-                // either be UserFacingError, or caught somewhere. However, we probably still want to catch arbitrary
-                // exceptions here just in case and return *some* reasonable response.
-                throw $e; // for now, so as to not break existing tests
+        if ($request is nonnull) {
```

Comment:
style nit: This would be a bit easier to read (and maybe safer) with early returns instead of these nested ifs

```Hack
try {
  // parse
} catch (...) {
  return ...
}

// no if needed here
try {
  // validate
} catch (...) {
  return ...
}

// no if needed
try {
  // resolve
} catch (...) {
  ...
}
```

Reply:
Yeah, I like that. Will update.